typescript/src/monitor-printers.ts

```diff
@@ -0,0 +1,120 @@
+    return lastFile && JSON.parse(await readFile(`${statusDir}/${lastFile}`, 'utf-8'));
+}
+
+/**
+ * Compares two lists of cluster statuses.
+ */
+function compareResults(previous: ClusterResponse[], current: ClusterResponse[], dateTime: Date): Comparison {
+    const getOnlineClusterIds = (clusters: ClusterResponse[]): Set<string> => new Set<string>(
+        clusters.filter((cluster) => cluster.is_online).map((cluster) => cluster.cluster_id),
+    );
+    const previousOnlineIds = getOnlineClusterIds(previous);
+    const currentOnlineIds = getOnlineClusterIds(current);
+    return {
+        dateTime: dateTime.toISOString(),
+        appeared: [...currentOnlineIds].filter((id) => !previousOnlineIds.has(id)),
+        gone: [...previousOnlineIds].filter((id) => !currentOnlineIds.has(id)),
+        onlineCount: currentOnlineIds.size,
+        offlineCount: current.length - currentOnlineIds.size,
+    };
+}
+
+/**
+ * Creates a CSV line from the given comparison.
+ */
+function csvLine(comparison: Comparison, separator = ';'): string {
+    const fields: (keyof Comparison)[] = ['dateTime', 'appeared', 'gone', 'onlineCount', 'offlineCount'];
+    return fields.map((field) => comparison[field]).join(separator);
+}
+
+/**
+ * Gets all clusters from the API periodically, writing the response to a JSON file & the comparisons to a CSV file.
+ */
+async function monitorClusters({
+    comparisonFile = '../cluster-monitoring-logs/cluster-monitoring.csv',
+    waitMS = 60000,
+    statusDir = '../cluster-monitoring-logs',
+}: MonitorClustersOptions = {}): Promise<void> {
+    let stop = false;
+    process.on('SIGINT', () => {
+        stop = true;
+    });
+
+    const demo = new DigitalFactoryDemo();
+    await demo.signIn();
+    print('Sign in completed.\n');
+
+    let previous = await getPreviousClusterStatus(statusDir);
+    const stream = createWriteStream(comparisonFile, { flags: 'a' });
+
+    try {
+        /* eslint-disable no-await-in-loop */
+        while (!stop) {
+            print('Retrieving clusters...');
+            const clusters = await demo.getClusters();
+            const dateTime = new Date();
+            const logFile = `${statusDir}/${dateTime.toISOString().replace(/:/g, '-')}-clusters.json`;
+            await writeFile(logFile, prettyJSON(clusters));
+            print(`Found ${clusters.length} clusters\n`);
+            if (previous) {
+                const comparison = compareResults(previous, clusters, dateTime);
+                stream.write(`${csvLine(comparison)}\n`);
+                print(`Comparison resulted in ${prettyJSON(comparison)}\n`);
+            }
+            previous = clusters;
+            await new Promise((resolve) => setTimeout(resolve, waitMS));
+        }
+    } catch (e: unknown) {
+        console.error(e); // eslint-disable-line no-console
+        process.exit(1);
+    } finally {
+        stream.end();
```

Comment:
I don't think this is actually going to work, after you exit the process on like 114, it won't make it to the `finally` statement.
![image](https://user-images.githubusercontent.com/88315443/145981361-7713b137-d27a-4a87-8189-07adae4f32e1.png)

Reply:
If the analysis and comparison part is moved out to different program, then this abort and exit stuff can just disappear.